go-controller/pkg/ovn/controller/podset/controller.go

```diff
@@ -0,0 +1,408 @@
+		psc.queue.AddRateLimited(key)
+	} else {
+		klog.Warningf("Dropping %q out of the queue: %v", key, err)
+		psc.queue.Forget(key)
+	}
+	utilruntime.HandleError(err)
+	return true
+}
+
+func (psc *PodSetController) processPod(namespace, name string) error {
+	pod, err := psc.podLister.Pods(namespace).Get(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return psc.handlePodDelete(namespace, name)
+		}
+
+		return err
+	}
+
+	return psc.handlePodUpdate(pod)
+}
+
+func (psc *PodSetController) processNamespace(name string) error {
+	ns, err := psc.namespaceLister.Get(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return psc.handleNamespaceDelete(name)
+		}
+	}
+	return psc.handleNamespaceUpdate(ns)
+}
+
+// AddPodSet adds an automatically-updating address set, based on one or
+// more selectors. This function is synchronous; it will block until the address set
+// is updated in nbdb.
+func (psc *PodSetController) AddPodSet(name string,
+	addressSet addressset.AddressSet,
+	selectors []Selector,
+) error {
+	psc.podSetsLock.Lock()
+	// no defer-unlock here, since we want to unlock
+	// before we sync
+
+	for _, selector := range selectors {
+		if selector.PodSelector == nil {
+			selector.PodSelector = labels.Everything()
+		}
+
+		if selector.NamespaceName == "" && selector.NamespaceSelector == nil {
+			// coding error, should never happen
+			panic("PodSet without namespace name or selector")
+		}
+	}
+
+	if len(selectors) == 0 {
+		klog.V(2).Infof("BUG: AddPodSet %s called with no selectors", name)
+		return nil
+	}
+
+	psc.podSets[name] = &PodSet{
+		name:       name,
+		addressSet: addressSet,
+
+		selectors: selectors,
+	}
+
+	klog.V(5).Infof("Adding PodSet %+v", psc.podSets[name])
+
+	// If the informers are still syncing, add this to the queue of
+	// pending podsets and exit.
+	if !(psc.podsSynced() && psc.namespacesSynced()) {
+		psc.podSetsToSync[name] = nil
+		psc.podSetsLock.Unlock()
+		return nil
+	}
+
+	// otherwise, sync the podset without holding the write lock (since that can be slow)
+	psc.podSetsLock.Unlock()
+
+	// but we're reading from psc.podSets, so we need the read lock
+	psc.podSetsLock.RLock()
+	defer psc.podSetsLock.RUnlock()
+	return psc.syncPodSet(psc.podSets[name])
+}
+
+func (psc *PodSetController) RemovePodSet(name string) {
+	psc.podSetsLock.Lock()
+	defer psc.podSetsLock.Unlock()
+
+	delete(psc.podSets, name)
+}
+
+// informer functions here
+
+func (psc *PodSetController) onPodAdd(obj interface{}) {
+	psc.enqueuePod(obj)
+}
+
+func (psc *PodSetController) onPodUpdate(old, new interface{}) {
+	oldPod := old.(*v1.Pod)
+	newPod := new.(*v1.Pod)
+
+	if shouldSkipPodUpdate(oldPod, newPod) {
+		return
+	}
+	psc.enqueuePod(new)
+}
+
+func (psc *PodSetController) onPodDelete(obj interface{}) {
+	klog.V(5).Info("PodSetController: Enqueueing pod delete")
+	psc.enqueuePod(obj)
```

Comment:
don't make the same mistake I did XD
https://github.com/ovn-org/ovn-kubernetes/pull/2122